content/en/blog/_posts/non-root-containers-and-devices.md

```diff
@@ -0,0 +1,196 @@
+that they are forbidden from running with root's group ID in `runAsGroup` or `supplementalGroups`.
+
+In Kubernetes*, kubelet builds the list of [`Device`](https://pkg.go.dev/k8s.io/cri-api@v0.22.1/pkg/apis/runtime/v1#Device)s to be made available to a container
+(based on inputs from the Device Plugins) and the list is included in the CreateContainer CRI message
+sent to the CRI container runtime. Each `Device` contains little information: host/container device
+paths and the desired devices cgroups permissions.
+
+The [OCI Runtime Spec for Linux Container Configuration](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md)
+expects that in addition to the devices cgroup fields, more detailed information about the devices
+must be provided:
+
+```json
+{
+        "type": "<string>",
+        "path": "<string>",
+        "major": <int64>,
+        "minor": <int64>,
+        "fileMode": <uint32>,
+        "uid": <uint32>,
+        "gid": <uint32>
+},
+```
+
+The CRI container runtimes (containerd, CRI-O) are responsible for obtaining this information
+from the host for each `Device`. By default, the runtimes copy the host device's user and group IDs:
+
+> `uid` (uint32, OPTIONAL) - id of device owner in the container namespace.  
+> `gid` (uint32, OPTIONAL) - id of device group in the container namespace.
+
+Similarly, the runtimes prepare other mandatory `config.json` sections based on the CRI fields,
+including the ones defined in `securityContext`: `runAsUser`/`runAsGroup`, which become part of the POSIX
+platforms user structure via:
+
+> `uid` (int, REQUIRED) specifies the user ID in the container namespace.  
+> `gid` (int, REQUIRED) specifies the group ID in the container namespace.  
+> `additionalGids` (array of ints, OPTIONAL) specifies additional group IDs in the container namespace to be added to the process.
+
+However, the resulting `config.json` triggers a problem when trying to run containers with
+both devices added and with non-root uid/gid set via `runAsUser`/`runAsGroup`: the container user process
+has no permission to use the device even when its group id (gid, copied from host) was permissive to
+non-root groups. This is because the container user does not belong to that host group (e.g., via `additionalGids`).
+
+Being able to run applications that use devices as non-root user is normal and expected to work so that
+the security principles can be met. Therefore, we looked into alternatives to get the gap filled with what the PodSec/CRI/OCI supports today.
+
+## What was done to solve the issue?
+Some readers probably noticed from the problem definition that it would at least be possible to workaround
+the problem by manually adding the device gid(s) to `supplementalGroups`, or in
+the case of just one device, set `runAsGroup` to the device's group id. However, this is problematic because the device gid(s) may have
+
+different values depending on the nodes' distro/version in the cluster:
+
+```bash
```

Comment:
Omit `bash` as this is not a valid script (it's a mix of commands, a prompt, and output).

Reply:
dropped